fix: TimeZone Normalisation

Fahad-Dezloper opened the pull request:
## What does this PR do?

Fixes an issue where busy times were not normalized to UTC, which caused incorrect availability calculations across time zones, particularly affecting Round Robin scheduling when team members are in different timezones.

The fix adds `.utc()` to the dayjs calls when formatting busy times, ensuring all times are compared in a consistent UTC format.

- Fixes #22150

## Updates since last revision

This PR was completed by Devin after becoming stale. Changes made:
- Rebased on upstream/main (file moved from `packages/lib/getUserAvailability.ts` to `packages/features/availability/lib/getUserAvailability.ts`)
- Applied the UTC normalization fix to the new file location
- Added comprehensive unit tests as requested by @zomars

### Tests Added
The new test file (`getUserAvailability.test.ts`) covers:
- Normalizing busy times from different timezones (IST, EST, GMT) to UTC
- Correctly identifying overlapping busy times when normalized
- Identifying no availability when all team members are busy at the same UTC time
- Daylight saving time transitions
- UTC+0 and UTC-0 format consistency
- Midnight crossing edge cases
- The original bug scenario from #22150

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). N/A - no documentation changes required.
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

Run the unit tests:
```bash
TZ=UTC yarn test packages/features/availability/lib/getUserAvailability.test.ts
```

All 7 tests should pass, verifying that busy times from different timezone formats are correctly normalized to UTC.

## Human Review Checklist

- [ ] Verify the fix is applied to the correct location (`packages/features/availability/lib/getUserAvailability.ts` lines 617-620)
- [ ] Confirm the unit tests adequately cover the timezone normalization scenarios for Round Robin scheduling
- [ ] Consider if additional integration tests are needed for multi-timezone team scenarios

---

## Summary by cubic
Fixed an issue where busy times were not normalized to UTC, which caused incorrect availability calculations across time zones.

---

Link to Devin run: https://app.devin.ai/sessions/e7ca2ccf39214eb993dc9f317202cc0b
Requested by: unknown ()

Co-authored-by: Fahad-Dezloper
<!-- devin-review-badge-begin -->

---

<a href="https://app.devin.ai/review/calcom/cal.com/pull/22162">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    <img src="https://static.devin.ai/assets/gh-open-in-devin-review-light.svg?v=1" alt="Open with Devin">
  </picture>
</a>
<!-- devin-review-badge-end -->

Devanshusharma2005 reviewed (dismissed):
LGTM

zomars commented on packages/lib/getUserAvailability.ts:
```diff
@@ -572,8 +572,8 @@ const _getUserAvailability = async function getUsersWorkingHoursLifeTheUniverseA
   const formattedBusyTimes = detailedBusyTimes.map((busy) => ({
-    start: dayjs(busy.start),
-    end: dayjs(busy.end),
+    start: dayjs(busy.start).utc(),
+    end: dayjs(busy.end).utc(),
```
@emrysal don't we have tests for catching these?

zomars reviewed (changes requested):
@Fahad-Dezloper  Thank you for your contribution 🙏 

Please add unit tests to prove this fix works and prevent regressions:

Expected Test Cases:
- `should normalize busy times from different timezones to UTC`
- `should correctly identify no availability when all team members are busy at the same UTC time`
- `should prevent the original bug where mixed timezone formats caused incorrect availability`
- Edge cases: daylight saving transitions, UTC+0/UTC-0 formats, invalid timezone data

These tests will ensure the Round Robin timezone bug doesn't reoccur and document the expected behavior.

Fahad-Dezloper commented:
sure will do

kart1ka commented:
@Fahad-Dezloper Please resolve the merge conflicts and add tests as suggested by zomars. Making it draft until then. Feel free to rfr once the comments are addressed.

devin-ai-integration[bot] committed:
fix: TimeZone Normalisation with unit tests

- Apply UTC normalization to busy times in getUserAvailability
- Add comprehensive unit tests for timezone normalization
- Tests cover: different timezone formats, DST transitions, UTC+0/-0 formats,
  midnight crossing, and the original bug scenario from #22150

Co-authored-by: Fahad-Dezloper <fahad@example.com>
Co-Authored-By: unknown <>

sahitya-chandra committed:
Merge branch 'main' into fix-TimeZone-Normalisation

sahitya-chandra committed:
Refactor timezone normalization tests for clarity

Refactor tests for timezone normalization in getUserAvailability, ensuring proper handling of busy times across different timezones and improving clarity in test cases.

sahitya-chandra reviewed (dismissed):
# Test Updated as per cubic review

**Problem:** Tests only asserted on `dayjs(...).utc()` output and didn’t verify the real availability behavior.

**Fix:** Tests now use the same path as production:
- Import `subtract` from `@calcom/features/schedules/lib/date-ranges`.
- Add `formatBusyTimesToUtc()` (same logic as in `getUserAvailability`).
- Call `subtract(dateRanges, formatBusyTimesToUtc(busyTimes))` and assert on the **returned availability** (which slots are free/blocked), not on raw dayjs values.

**Result:** The same 7 scenarios now validate timezone normalization via the real `subtract()` flow instead of only `dayjs.utc()`.

```bash
TZ=UTC yarn test packages/features/availability/lib/getUserAvailability.test.ts --run 2>&1
```

<img width="1183" height="337" alt="Image" src="https://github.com/user-attachments/assets/aa4e4a61-7292-455d-8232-2711c2ef7c14" />

sahitya-chandra committed:
fix: use real formatBusyTimesToUtc in tests, exercise subtract() path

- Export formatBusyTimesToUtc from getUserAvailability; use it in production
- Tests import formatBusyTimesToUtc from production instead of duplicating logic
- Tests assert on subtract() result (availability) instead of raw dayjs.utc()
- Addresses P2: no drift if normalization logic changes

Fahad-Dezloper committed:
Merge branch 'main' into fix-TimeZone-Normalisation
